File: clients/rust/src/instructions/initialize_oracle/initialize_oracle.rs
use crate::{
    derive_controller_authority_pda, derive_oracle_pda,
    generated::instructions::InitializeOracleBuilder,
};
use solana_instruction::Instruction;
use solana_program::system_program;
use solana_pubkey::Pubkey;

pub fn create_initialize_oracle_instruction(
    controller: &Pubkey,
    authority: &Pubkey,
    nonce: &Pubkey,
    price_feed: &Pubkey,
    oracle_type: u8,
    mint: &Pubkey,
    quote_mint: &Pubkey,
) -> Instruction {
    let controller_authority = derive_controller_authority_pda(controller);
    let oracle_pda = derive_oracle_pda(&nonce);
    InitializeOracleBuilder::new()
        .controller(*controller)
        .controller_authority(controller_authority)
        .authority(*authority)
        .oracle(oracle_pda)
        .price_feed(*price_feed)
        .system_program(system_program::ID)
        .payer(*authority)
        .oracle_type(oracle_type)
        .nonce(*nonce)
        .base_mint(*mint)
        .quote_mint(*quote_mint)
        .instruction()
}

File: clients/ts/src/instructions/initializeOracle.ts
import { Address, TransactionSigner } from "@solana/kit";
import { getInitializeOracleInstruction } from "../generated/instructions/initializeOracle";
import { deriveControllerAuthorityPda, deriveOraclePda } from "../pdas";
import { SVM_ALM_CONTROLLER_PROGRAM_ADDRESS } from "../generated";

const SYSTEM_PROGRAM_ID = "11111111111111111111111111111111" as Address;

/**
 * Instruction generation for initializing an oracle account
 */
export async function createInitializeOracleInstruction(
  controller: Address,
  authority: TransactionSigner,
  nonce: Address,
  priceFeed: Address,
  oracleType: number,
  mint: Address,
  quoteMint: Address
) {
  const controllerAuthority = await deriveControllerAuthorityPda(controller);
  const oraclePda = await deriveOraclePda(nonce);

  return getInitializeOracleInstruction({
    controller,
    controllerAuthority,
    authority,
    oracle: oraclePda,
    priceFeed,
    payer: authority,
    systemProgram: SYSTEM_PROGRAM_ID,
    oracleType,
    nonce,
    baseMint: mint,
    quoteMint,
  });
}

